### PR TITLE
#12278 remove existing extension attributes during extraction of exte…

### PR DESCRIPTION
**Pull request by Zefiryn**
…nsion attribues when existing class does not extend correct interface

When extracting extension attribute from data of one class make sure the result will not have extension attributes defined for that class if the two entities do not share same extension class. Otherwise it may result in error when trying copy one class data to another.

### Description
Changes are designed to handle the case when input `$data` contains `extension_attributes` field but the class for which extension attributes need to be extracted does not have any extension attributes defined. In such case it is possible that extension attribute from source object will cause type error when passed to constructor of the `$extensibleEntityClass`

### Fixed Issues (if relevant)
1. magento/magento2#12278: Error while fetching bundle item with repository

### Manual testing scenarios
1. Create a bundle product with at least 1 option of simple product
2. Create a module which defines any extension attributes for `Magento\Catalog\Api\Data\ProductInterface`
3. Create an observer for `catalog_product_collection_load_after` event and in that observer set extension attribute defined in step 2 on each item
4. Try to retrieve bundle product using `Magento\Catalog\Api\ProductRepositoryInterface::getById()` method providing only id of the bundle product created in step 1. 
5. No error should be raised. 

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### Ports to other versions
* `2.1-develop` https://github.com/magento/magento2/pull/12280
* `2.3-develop` https://github.com/magento/magento2/pull/12281

### CODE REVIEW

**Comment by dmanners** [comment]
Hello @Zefiryn, sorry for the delay in getting to this pull request. There are a few failures in the travis test suite. Could you please have a look into these before I process this pull request.

Thanks.

**Comment by Zefiryn** [comment]
@dmanners sure, I will take a look at it during the weekend

**Comment by Zefiryn** [comment]
@dmanners I have checked the errors and decided to redo the approach and to only fix the problem with loading bundle product originally posted in the issue. I noticed that several tests uses regular arrays when setting `extension_attributes` data in the object. I am not sure why this has been done this way as `extension_attributes` is always an object. Anyway updated PRs for branches 2.1-develop, 2.2-develop and 2.3-develop are ready.

File: app/code/Magento/Bundle/Model/Product/LinksList.php [hunk 1]
```diff
@@ -70,4 +77,20 @@ public function getItems(\Magento\Catalog\Api\Data\ProductInterface $product, $o
...
+     * Get selection product data and remove extension attributes if necessary
+     *
+     * @param ProductInterface $selection
+     * @return array
+     */
+    private function getSelectionData(ProductInterface $selection)
+    {
+        $selectionData = $selection->getData();
+        if (array_key_exists(ExtensibleDataInterface::EXTENSION_ATTRIBUTES_KEY, $selectionData)) {
+            unset($selectionData[ExtensibleDataInterface::EXTENSION_ATTRIBUTES_KEY]);
```
Comment (orlangur) [inline app/code/Magento/Bundle/Model/Product/LinksList.php:91]: Please replace with unconditional `unset`.

File: lib/internal/Magento/Framework/Api/ExtensionAttribute/JoinProcessor.php [hunk 2]
```diff
@@ -100,7 +101,10 @@ private function getReferenceTableAlias($attributeCode)
     public function extractExtensionAttributes($extensibleEntityClass, array $data)
     {
         if (!$this->isExtensibleAttributesImplemented($extensibleEntityClass)) {
-            /* do nothing as there are no extension attributes */
+            /* remove any extension attributes from source data as destination entity has no extension attributes */
+            if (array_key_exists(ExtensibleDataInterface::EXTENSION_ATTRIBUTES_KEY, $data)) {
+                unset($data[ExtensibleDataInterface::EXTENSION_ATTRIBUTES_KEY]);
```
Comment (orlangur) [inline lib/internal/Magento/Framework/Api/ExtensionAttribute/JoinProcessor.php:106]: Please replace with unconditional `unset`.

File: app/code/Magento/Bundle/Model/Product/LinksList.php [hunk 3]
```diff
@@ -70,4 +77,20 @@ public function getItems(\Magento\Catalog\Api\Data\ProductInterface $product, $o
...
         return $productLinks;
     }
+
+    /**
+     * Get selection product data and remove extension attributes if necessary
+     *
+     * @param ProductInterface $selection
+     * @return array
+     */
+    private function getSelectionData(ProductInterface $selection)
```
Comment (orlangur) [inline app/code/Magento/Bundle/Model/Product/LinksList.php:87]: Please rename to `getSelectionDataWithoutExtensionAttributes`.

**Review by orlangur — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
Hi @Zefiryn,
I'm closing this Pull Request due to inactivity. Feel free to fix all requested changes and reopen your PR or create new one.

**Comment by sheepfy** [comment]
@ihor-sviziev 
Hei, is this up-to-date? I mean i currently have the same issue with the extension attributes.
You just cannot use both extension_attributes and bundle products at the same time. 

This PR seems to fix this issue. Is it OK if i continue with this PR ?

**Pull request closed without merging**